truncate_client/src/app_inner.rs

```diff
@@ -141,29 +144,93 @@ pub fn render(client: &mut OuterApplication, ui: &mut egui::Ui) {
             }
         }
         GameStatus::Tutorial(tutorial) => {
-            tutorial.render(ui, theme);
+            tutorial.render(ui, theme, current_time);
         }
         GameStatus::SinglePlayer(sp) => {
-            sp.render(ui, theme);
+            sp.render(ui, theme, current_time);
         }
         GameStatus::PendingJoin(room_code) => {
             ui.label(format!("Waiting to join room {room_code}"));
+
+            let dot_count = (current_time.as_millis() / 500) % 4;
+            let mut dots = vec!["."; dot_count as usize];
+            dots.extend(vec![" "; 4 - dot_count as usize]);
+            let msg = format!("JOINING {room_code}{}", dots.join(""));
+
+            let msg_text = TextHelper::heavy(&msg, 14.0, ui);
+            let button_text = TextHelper::heavy("CANCEL", 14.0, ui);
+            let required_size = vec2(
+                msg_text.size().x,
+                msg_text.size().y + button_text.size().y * 2.0,
+            );
+
+            let margins = (ui.available_size_before_wrap() - required_size) / 2.0;
+            let outer_frame = egui::Frame::none().inner_margin(Margin::from(margins));
+            outer_frame.show(ui, |ui| {
+                msg_text.paint(Color32::WHITE, ui);
+                ui.add_space(8.0);
+                if button_text
+                    .button(
+                        theme.selection.lighten().lighten(),
+                        theme.text,
+                        &map_texture,
+                        ui,
+                    )
+                    .clicked()
+                {
+                    #[cfg(target_arch = "wasm32")]
+                    {
+                        web_sys::window().unwrap().location().reload();
+                    }
+                }
+            });
         }
         GameStatus::PendingCreate => {
-            ui.label("Waiting for a new game to be created . . .");
+            let dot_count = (current_time.as_millis() / 500) % 4;
```

Comment:
Code between `PendingJoin` and `PendingCreate` is basically duplicated yeah? Nbd though

Reply:
For now yeah. I'd like to move it all the way out to be some nice "dottable" component or something but not right now